gazebo_plugins/src/gazebo_ros_diff_drive.cpp

```diff
@@ -350,21 +380,83 @@ void GazeboRosDiffDrive::QueueThread()
 
 void GazeboRosDiffDrive::UpdateOdometryEncoder()
 {
-    double vl = joints_[LEFT]->GetVelocity ( 0 );
-    double vr = joints_[RIGHT]->GetVelocity ( 0 );
 #if GAZEBO_MAJOR_VERSION >= 8
     common::Time current_time = parent->GetWorld()->SimTime();
 #else
     common::Time current_time = parent->GetWorld()->GetSimTime();
 #endif
+
+    // Handle dropouts and delays
+    bool use_old_data = !is_reading_;
+
+    // Handle dropout update
+    if((dropout_set_ != DropoutSet::DROPOUT_NONE) || (!is_reading_))
+    {
+        // If dropouts are defined or a delay occurred (so no read currently)
+        if((current_time - last_dropout_change_).Double() > next_dropout_change_s_)
+        {
+            if(is_reading_)
+            {
+                // Going to not reading - need to see if we turn off any future dropouts
+                if(dropout_set_ == DropoutSet::DROPOUT_ONCE)
+                {
+                    // By doing the change this way, a delay and a single dropout still functions correctly.
+                    dropout_set_ = DropoutSet::DROPOUT_NONE;
+                }
+                // Already know a dropout should occur in this if statement
+                // Calculate the length of the dropout
+                next_dropout_change_s_ = (dropout_length_max_s_ - dropout_length_min_s_) *
+                    ignition::math::Rand::DblUniform() + dropout_length_min_s_;
+            }
+            else if(dropout_set_ != DropoutSet::DROPOUT_NONE)
+            {
+                // Currently is not reading and will have another dropout
+                next_dropout_change_s_ = (dropout_delay_max_s_ - dropout_delay_min_s_) *
+                    ignition::math::Rand::DblUniform() + dropout_delay_min_s_;
+            }
+            // else: is not reading, but no more dropouts - just go to reading without anything further.
+
+            // Make the change
+            is_reading_ = !is_reading_;
+            ROS_DEBUG_STREAM("Encoder is reading: " << is_reading_);
+            // Reset the last dropout change
+            last_dropout_change_ = current_time;
+        }
+    }
+    // Once starts reading, this becomes false.
+    if(is_reading_)
+    {
+        is_delayed_start_ = false;
+    }
+
+    double vl = joints_[LEFT]->GetVelocity ( 0 );
+    double vr = joints_[RIGHT]->GetVelocity ( 0 );
+    if(use_old_data)
+    {
+        // No encoder update, so use last known data (extrapolate)
+        vl = last_vl_;
+        vr = last_vr_;
+    }
```

Comment:
So during a dropout the wheel speeds are constant and not necessarily zero?

Reply:
After our Slack discussion, I updated the behavior to a 0 speed signal when the encoder drops out. The failure is at a different point in the chain than I was first replicating.